test/validator/insert_validator.js

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+const BN = require('bn.js');
+const { AccountProvider } = require('../test_lib/utils.js');
+const Utils = require('../test_lib/utils.js');
+
+const ValidatorSet = artifacts.require('TestValidatorSet');
+
+contract('ValidatorSet::insertValidator', (accounts) => {
+  const accountProvider = new AccountProvider(accounts);
+  let validatorSet;
+  const beginHeight = new BN(100);
+  beforeEach(async () => {
+    validatorSet = await ValidatorSet.new();
+  });
+
+  contract('Positive Tests', () => {
+    it('should insert validators', async () => {
+      const account1 = accountProvider.get();
+      const account2 = accountProvider.get();
+
+      await validatorSet.addValidator(account1, beginHeight);
+
+      const actualValidatorBeginHeight = await validatorSet.validatorBeginHeight.call(account1);
+      const actualValidatorEndHeight = await validatorSet.validatorEndHeight.call(account1);
+      assert.strictEqual(
+        beginHeight.eq(actualValidatorBeginHeight),
+        true,
+        `Expected validator begin height is ${beginHeight} but got ${actualValidatorBeginHeight}`,
+      );
+
+      assert.strictEqual(
+        actualValidatorEndHeight.toString(16),
+        Utils.MAX_FUTURE_END_HEIGHT,
+        `Expected validator end height is ${Utils.MAX_FUTURE_END_HEIGHT} `
+         + `but got ${actualValidatorEndHeight}`,
+      );
+
+      // Inserting another validator.
+      await validatorSet.addValidator(account2, beginHeight);
```

Comment:
This is internal details how we add validators. I would rather propose to check that it exists in linked list.

Reply:
I think it checks whether the validator is present in the linked list.
Are you suggesting to traverse the linked list, get the validator adresses and then validate against the expected list of validators ?

Reply:
To check if validator presents in the linked list it should not point to null address, yes?

Reply:
New added validator is pointed by **SENTINEL_ADDRESS**[0x1]. So in the test case, 2nd validator added is pointed by 0x1.
Example :
1. Consider we add validator `A`. After addition the validators mapping will be :
```
[0x1] = A
[A] = 0x1
```
2. After adding validator `B`, the validators mapping would be :
```
[0x1] = B
[B] = A
[A] = 0x1
```
In the test case, below validations are done :
1. At `0x1` i.e SENTINEL_ADDDRESS, validator `B` is present.
2. At `B` address, validator `A` is present.

Reply:
> To check if validator presents in the linked list it should not point to null address, yes?
correct, existence check suffices that it is not-zero

Reply:
@pgev , I did the suggested changes.